lib/puppet/data_providers/lookup_adapter.rb

```diff
@@ -0,0 +1,199 @@
+    options = nil
+    unless module_name.nil?
+      catch(:no_such_key) do
+        options = module_provider(module_name).lookup(LOOKUP_OPTIONS, invocation, merge_strategy)
+        options = merge.merge(env_opts, options) unless env_opts.nil?
+      end
+    end
+    options.nil? ? env_opts : options
+  end
+
+  # Retrieve and cache lookup options specific to the environment that this adapter is attached to (i.e. a merge
+  # of global and environment lookup options).
+  def env_lookup_options(lookup_invocation, merge_strategy)
+    unless instance_variable_defined?(:@env_lookup_options)
+      @env_lookup_options = nil
+      catch(:no_such_key) do
+        @env_lookup_options = merge_strategy.merge_lookup([:lookup_global, :lookup_in_environment]) do |m|
+          send(m, LOOKUP_OPTIONS, lookup_invocation, merge_strategy)
+        end
+      end
+    end
+    @env_lookup_options
+  end
+
+  def env_provider
+    @env_provider ||= initialize_env_provider
+  end
+
+  def module_provider(module_name)
+    # Test if the key is present for the given module_name. It might be there even if the
+    # value is nil (which indicates that no module provider is configured for the given name)
+    unless data.include?(module_name)
+      data[module_name] = initialize_module_provider(module_name)
+    end
+    data[module_name]
+  end
+
+  def initialize_module_provider(module_name)
+    injector = Puppet.lookup(:injector) { nil }
+
+    # Support running tests without an injector being configured == using a null implementation
+    return ModuleDataProvider.new() unless injector
+
+    # Get the registry of module to provider implementation name
+    module_service_type = Registry.hash_of_per_module_data_provider
+    module_service_name = PER_MODULE_DATA_PROVIDER_KEY
+    module_service = injector.lookup(nil, module_service_type, module_service_name)
+    provider_name = module_service[module_name] || 'none'
+
+    service_type = Registry.hash_of_module_data_providers
+    service_name = MODULE_DATA_PROVIDERS_KEY
+
+    # Get the service (registry of known implementations)
+    service = injector.lookup(nil, service_type, service_name)
+    provider = service[provider_name]
+    unless provider
+      raise Puppet::Error.new("Environment '#{@env.name}', cannot find module_data_provider '#{provider_name}'")
+    end
+    provider
+  end
+
+  def initialize_env_provider
+    injector = Puppet.lookup(:injector) { nil }
+
+    # Support running tests without an injector being configured == using a null implementation
+    return EnvironmentDataProvider.new() unless injector
+
+    # Get the environment's configuration since we need to know which data provider
+    # should be used (includes 'none' which gets a null implementation).
+    #
+    env_conf = @env.configuration
+
+    # Get the data provider and find the bound implementation
+    # TODO: PUP-1640, drop the nil check when legacy env support is dropped
```

Comment:
legacy env support has been dropped in puppet 4.x - there is always an env.conf now.